- Buat system rekomendasi content based dan collaborative filtering
- Tentukan daftar rekomendasi anime untuk:
    - Content based berdasarkan ```Genre```:
        - user 'Ali' baca anime no 1
        - user 'Bebi' baca anime no 2,24,25
        - user 'Cindy' baca anime no 10,200
        - user dengan ```user id = 1 ```
    - Collaborative:
        - user 'Ali' memberikan rating no 1:10
        - user 'Bebi' memberikan rating no 2: 8, no 24:-1, no 25:5
        - user 'Cindy' memberikan rating no 10:10, no 200:1
        - user dengan ```user id = 1 ```

# Content Based

In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import CountVectorizer

In [2]:
df = pd.read_csv('anime.csv')
# df = df.iloc[:1000]
df2 = pd.read_csv('rating.csv')
df2.head()

user_id  anime_id  rating
0        1        20      -1
1        1        24      -1
2        1        79      -1
3        1       226      -1
4        1       241      -1

In [9]:
df.head(15)

anime_id                                               name  \
0      32281                                     Kimi no Na wa.   
1       5114                   Fullmetal Alchemist: Brotherhood   
2      28977                                           Gintama°   
3       9253                                        Steins;Gate   
4       9969                                      Gintama&#039;   
5      32935  Haikyuu!!: Karasuno Koukou VS Shiratorizawa Ga...   
6      11061                             Hunter x Hunter (2011)   
7        820                               Ginga Eiyuu Densetsu   
8      15335  Gintama Movie: Kanketsu-hen - Yorozuya yo Eien...   
9      15417                           Gintama&#039;: Enchousen   
10      4181                               Clannad: After Story   
11     28851                                     Koe no Katachi   
12       918                                            Gintama   
13      2904                 Code Geass: Hangyaku no Lelouch R2   
14     28891                            Haikyuu!! Second Season   

                                                genre   type episodes  rating  \
0                Drama, Romance, School, Supernatural  Movie        1    9.37   
1   Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64    9.26   
2   Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.25   
3                                    Sci-Fi, Thriller     TV       24    9.17   
4   Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.16   
5              Comedy, Drama, School, Shounen, Sports     TV       10    9.15   
6             Action, Adventure, Shounen, Super Power     TV      148    9.13   
7                      Drama, Military, Sci-Fi, Space    OVA      110    9.11   
8   Action, Comedy, Historical, Parody, Samurai, S...  Movie        1    9.10   
9   Action, Comedy, Historical, Parody, Samurai, S...     TV       13    9.11   
10  Drama, Fantasy, Romance, Slice of Life, Supern...     TV       24    9.06   
11                             Drama, School, Shounen  Movie        1    9.05   
12  Action, Comedy, Historical, Parody, Samurai, S...     TV      201    9.04   
13  Action, Drama, Mecha, Military, Sci-Fi, Super ...     TV       25    8.98   
14             Comedy, Drama, School, Shounen, Sports     TV       25    8.93   

    members  
0    200630  
1    793665  
2    114262  
3    673572  
4    151266  
5     93351  
6    425855  
7     80679  
8     72534  
9     81109  
10   456749  
11   102733  
12   336376  
13   572888  
14   179342

In [3]:
df.isnull().sum()

anime_id      0
name          0
genre        62
type         25
episodes      0
rating      230
members       0
dtype: int64

In [4]:
df = df.dropna()

df.isnull().sum()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12017 entries, 0 to 12293
Data columns (total 7 columns):
anime_id    12017 non-null int64
name        12017 non-null object
genre       12017 non-null object
type        12017 non-null object
episodes    12017 non-null object
rating      12017 non-null float64
members     12017 non-null int64
dtypes: float64(1), int64(2), object(4)
memory usage: 751.1+ KB


In [5]:
cv = CountVectorizer(tokenizer= lambda x: x.split(', '))
ca = cv.fit_transform(df['genre'])
ca = ca.toarray()
print(cv.get_feature_names())


# reduce float precision
# ca = ca.astype(np.int32)
ca

['action', 'adventure', 'cars', 'comedy', 'dementia', 'demons', 'drama', 'ecchi', 'fantasy', 'game', 'harem', 'hentai', 'historical', 'horror', 'josei', 'kids', 'magic', 'martial arts', 'mecha', 'military', 'music', 'mystery', 'parody', 'police', 'psychological', 'romance', 'samurai', 'school', 'sci-fi', 'seinen', 'shoujo', 'shoujo ai', 'shounen', 'shounen ai', 'slice of life', 'space', 'sports', 'super power', 'supernatural', 'thriller', 'vampire', 'yaoi', 'yuri']


array([[0, 0, 0, ..., 0, 0, 0],
       [1, 1, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [6]:
# 3. Cosine similarity dari tiap data di matrix freq

cosScore = cosine_similarity(ca)
cosScore

array([[1.        , 0.18898224, 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.18898224, 1.        , 0.28571429, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.28571429, 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 1.        ,
        1.        ],
       [0.        , 0.        , 0.        , ..., 1.        , 1.        ,
        1.        ],
       [0.        , 0.        , 0.        , ..., 1.        , 1.        ,
        1.        ]])

## Joblib

In [10]:
import joblib

In [11]:
joblib.dump(cosScore, 'Rekomendasi')

['Rekomendasi']

### Rekomendasi Ali

In [13]:
Ali = df[df['anime_id'] == 1].index[0]

similarAli= list(enumerate(cosScore[Ali]))
# similarAli # (index, %similarity)

In [14]:
similarAli = sorted(similarAli, key=lambda x: x[1], reverse=True)
# similarAli

In [15]:
indexAli = []
for i in similarAli[:11]:
    x = i[0]
    indexAli.append(x)
indexAli

[22, 1465, 6565, 551, 1073, 1398, 2333, 2735, 3056, 5719, 700]

In [16]:
judulAli = []
for i in indexAli:
    x = df['name'].iloc[i]
    judulAli.append(x)
judulAli

['Cowboy Bebop',
 'Cowboy Bebop: Yose Atsume Blues',
 'Seihou Tenshi Angel Links',
 'Seihou Bukyou Outlaw Star',
 'Waga Seishun no Arcadia',
 'Waga Seishun no Arcadia: Mugen Kidou SSX',
 'Ginga Tetsudou Monogatari',
 'Uchuu Kaizoku Captain Harlock: Arcadia-gou no Nazo',
 'Ginga Tetsudou Monogatari: Eien e no Bunkiten',
 'Kaitei Choutokkyuu: Marine Express',
 'Urusei Yatsura Movie 2: Beautiful Dreamer']

In [17]:
dfAli = []
for i in similarAli[:11]:
    dfAli.append(df.iloc[i[0]])
dfAli = pd.DataFrame(dfAli)
dfAli

anime_id                                               name  \
22           1                                       Cowboy Bebop   
1465      4037                    Cowboy Bebop: Yose Atsume Blues   
6568      1226                          Seihou Tenshi Angel Links   
551        400                          Seihou Bukyou Outlaw Star   
1073      2202                            Waga Seishun no Arcadia   
1398      2203           Waga Seishun no Arcadia: Mugen Kidou SSX   
2333      1490                          Ginga Tetsudou Monogatari   
2735      2470  Uchuu Kaizoku Captain Harlock: Arcadia-gou no ...   
3057      2717      Ginga Tetsudou Monogatari: Eien e no Bunkiten   
5721      3175                 Kaitei Choutokkyuu: Marine Express   
700       1921          Urusei Yatsura Movie 2: Beautiful Dreamer   

                                                  genre     type episodes  \
22      Action, Adventure, Comedy, Drama, Sci-Fi, Space       TV       26   
1465    Action, Adventure, Comedy, Drama, Sci-Fi, Space  Special        1   
6568  Action, Adventure, Comedy, Drama, Romance, Sci...       TV       13   
551            Action, Adventure, Comedy, Sci-Fi, Space       TV       24   
1073            Action, Adventure, Drama, Sci-Fi, Space    Movie        1   
1398            Action, Adventure, Drama, Sci-Fi, Space       TV       22   
2333            Action, Adventure, Drama, Sci-Fi, Space       TV       26   
2735            Action, Adventure, Drama, Sci-Fi, Space    Movie        1   
3057            Action, Adventure, Drama, Sci-Fi, Space       TV       24   
5721           Action, Adventure, Comedy, Drama, Sci-Fi  Special        1   
700   Action, Adventure, Comedy, Drama, Romance, Sci-Fi    Movie        1   

      rating  members  
22      8.82   486824  
1465    7.53    19565  
6568    6.06     4817  
551     7.98    78600  
1073    7.70     4415  
1398    7.56     2587  
2333    7.29     5947  
2735    7.18     2579  
3057    7.09     2072  
5721    6.37      551  
700     7.89     5663

### Rekomendasi Bebi

In [50]:
Bebi = df[df['anime_id'] == 25].index
# print(Bebi) # 0,464,1432

similarBebi1= list(enumerate(cosScore[0]))
similarBebi2= list(enumerate(cosScore[464]))
similarBebi3= list(enumerate(cosScore[1432]))

In [51]:
similarBebi1 = sorted(similarBebi1, key=lambda x: x[1], reverse=True)
similarBebi2 = sorted(similarBebi2, key=lambda x: x[1], reverse=True)
similarBebi3 = sorted(similarBebi3, key=lambda x: x[1], reverse=True)


In [52]:
dfBebi = []
for i in similarBebi1[:4],similarBebi2[:4],similarBebi3[:4]:
    dfBebi.append(df.iloc[i[0][0]])
    dfBebi.append(df.iloc[i[1][0]])
    dfBebi.append(df.iloc[i[2][0]])
dfBebi = pd.DataFrame(dfBebi)
dfBebi

anime_id                                name  \
0        32281                      Kimi no Na wa.   
5805       547         Wind: A Breath of Heart OVA   
6394       546        Wind: A Breath of Heart (TV)   
367        846              School Rumble Ni Gakki   
464         24                       School Rumble   
772        517     School Rumble Ichi Gakki Hoshuu   
1432        25                           Sunabouzu   
4486       370                      Burn Up Excess   
2760     33713  Seihou Bukyou Outlaw Star Specials   

                                                  genre     type episodes  \
0                  Drama, Romance, School, Supernatural    Movie        1   
5805               Drama, Romance, School, Supernatural      OVA        3   
6394               Drama, Romance, School, Supernatural       TV       13   
367                    Comedy, Romance, School, Shounen       TV       26   
464                    Comedy, Romance, School, Shounen       TV       26   
772                    Comedy, Romance, School, Shounen      OVA        2   
1432  Action, Adventure, Comedy, Ecchi, Sci-Fi, Shounen       TV       24   
4486  Action, Adventure, Comedy, Ecchi, Police, Sci-...       TV       13   
2760           Action, Adventure, Comedy, Ecchi, Sci-Fi  Special        2   

      rating  members  
0       9.37   200630  
5805    6.35     2043  
6394    6.14     7778  
367     8.15    93095  
464     8.06   178553  
772     7.85    40868  
1432    7.55    65516  
4486    6.71     6402  
2760    7.16     1509

### Rekomendasi Cindy

In [87]:
## anime id = 100 dan 200
Cindy1 = df[df['anime_id']==100].index[0] #2210
Cindy2 = df[df['anime_id']==200].index[0] #1302

similarCindy1= list(enumerate(cosScore[Cindy1]))
similarCindy2= list(enumerate(cosScore[Cindy2]))
similarCindy1 = sorted(similarCindy1, key = lambda i: i[1], reverse=True)
similarCindy2 = sorted(similarCindy2, key = lambda i: i[1], reverse=True)
# similarCindy1

In [88]:
# index1 = []
# index2 = []
# for i in similarCindy1[:5]:
#     index1.append(i[0])
# for i in similarCindy2[:5]:
#     index2.append(i[0])
# index1,index2
    

([2210, 233, 534, 925, 1399], [1302, 4160, 4752, 184, 4403])

In [90]:
Cindy = similarCindy1+similarCindy2
Cindy = sorted(Cindy, key = lambda i: i[1], reverse=True)
# Cindy

In [94]:
index = []
for i in Cindy[:10]:
    index.append(i[0])
Cindy = df.iloc[index]
Cindy

anime_id                                               name  \
1302       200                                Tenshi na Konamaiki   
2210       100               Shin Shirayuki-hime Densetsu Prétear   
233        372       Cardcaptor Sakura Movie 2: Fuuin Sareta Card   
4162       626                               Mamotte Shugogetten!   
4754       660                        Denshin Mamotte Shugogetten   
184       7674                                           Bakuman.   
4405       590                                  Kage kara Mamoru!   
7034      2081                                 Makeruna! Makendou   
534        996          Bishoujo Senshi Sailor Moon: Sailor Stars   
925        997  Bishoujo Senshi Sailor Moon S: Kaguya Hime no ...   

                                                  genre   type episodes  \
1302                    Comedy, Magic, Romance, Shounen     TV       50   
2210  Comedy, Drama, Fantasy, Magic, Romance, Shoujo...     TV       13   
233      Comedy, Drama, Fantasy, Magic, Romance, Shoujo  Movie        1   
4162           Comedy, Fantasy, Magic, Romance, Shounen     TV       22   
4754           Comedy, Fantasy, Magic, Romance, Shounen    OVA        8   
184                            Comedy, Romance, Shounen     TV       25   
4405                           Comedy, Romance, Shounen     TV       12   
7034                             Comedy, Magic, Shounen    OVA        1   
534   Adventure, Comedy, Drama, Fantasy, Magic, Roma...     TV       34   
925   Adventure, Comedy, Drama, Fantasy, Magic, Roma...  Movie        1   

      rating  members  
1302    7.60    15077  
2210    7.33    37394  
233     8.28    60637  
4162    6.79     4995  
4754    6.64     2372  
184     8.35   246899  
4405    6.73    17136  
7034    5.78      555  
534     7.99    52586  
925     7.76    29394

### Rekomendasi User ID = 1

In [120]:
df2.head()

user_id  anime_id  rating
0        1        20      -1
1        1        24      -1
2        1        79      -1
3        1       226      -1
4        1       241      -1

In [127]:
Dedi = df2[df2['user_id']==1].sort_values(by='rating',ascending=False)
Dedi

user_id  anime_id  rating
47         1      8074      10
81         1     11617      10
101        1     15451      10
83         1     11757      10
106        1     16706      -1
..       ...       ...     ...
52         1      8841      -1
53         1      9041      -1
54         1      9062      -1
55         1      9136      -1
152        1     31845      -1

[153 rows x 3 columns]

User ID=1 memberikan rating 10 untuk 4 anime_id 8074,11617,15451, dan 11757, selain itu usir id 1 memberi rating -1 atau artinya dia menonton tapi tidak memberikan rating.

In [128]:
Dedi1 = Dedi['anime_id'].iloc[:4].index
Dedi1[0]

47

In [132]:
similarDedi1= list(enumerate(cosScore[Dedi1[0]]))
similarDedi2= list(enumerate(cosScore[Dedi1[1]]))
similarDedi3= list(enumerate(cosScore[Dedi1[2]]))
similarDedi4= list(enumerate(cosScore[Dedi1[3]]))

In [133]:
similarDedi = similarDedi1 + similarDedi2 + similarDedi3 + similarDedi4
similarDedi = sorted(similarDedi, key = lambda i: i[1], reverse=True)
similarDedi[:10]

[(47, 1.0000000000000002),
 (149, 1.0000000000000002),
 (83, 1.0000000000000002),
 (3029, 1.0000000000000002),
 (101, 0.9999999999999999),
 (268, 0.9999999999999999),
 (290, 0.9999999999999999),
 (461, 0.9999999999999999),
 (795, 0.9999999999999999),
 (3831, 0.9999999999999999)]

In [136]:
index = []
for i in similarDedi[:20]:
    index.append(i[0])
dfDedi = df.iloc[index]
dfDedi

anime_id                                               name  \
47       22135                            Ping Pong The Animation   
149       5040                                           One Outs   
83         877                                               Nana   
3030      1869                                        Nana Recaps   
101      18115                         Magi: The Kingdom of Magic   
268      14513                       Magi: The Labyrinth of Magic   
290      31741                        Magi: Sinbad no Bouken (TV)   
461      22097                             Magi: Sinbad no Bouken   
795       8086                     Densetsu no Yuusha no Densetsu   
3833     21235                                  Log Horizon Recap   
4235      7436  Dragon Quest: Dai no Daibouken Tachiagare!! Ab...   
4445      7364  Dragon Quest: Dai no Daibouken Buchiyabure!! S...   
4978      9690        Densetsu no Yuusha no Densetsu: Iris Report   
6306     25431                   Chain Chronicle: Short Animation   
6785      2782                                        Fire Emblem   
8298     33728       Chain Chronicle: Haecceitas no Hikari Part 1   
9529     18311                                        Meoteoldosa   
9531     18329                               Meoteoldosawa Ttomae   
81       18617                          Girls und Panzer der Film   
932      18619  Girls und Panzer: Kore ga Hontou no Anzio-sen ...   

                                                  genre     type episodes  \
47                        Psychological, Seinen, Sports       TV       11   
149                       Psychological, Seinen, Sports       TV       25   
83    Comedy, Drama, Music, Romance, Shoujo, Slice o...       TV       47   
3030  Comedy, Drama, Music, Romance, Shoujo, Slice o...  Special        3   
101          Action, Adventure, Fantasy, Magic, Shounen       TV       25   
268          Action, Adventure, Fantasy, Magic, Shounen       TV       25   
290          Action, Adventure, Fantasy, Magic, Shounen       TV       13   
461          Action, Adventure, Fantasy, Magic, Shounen      OVA        5   
795          Action, Adventure, Fantasy, Magic, Shounen       TV       24   
3833         Action, Adventure, Fantasy, Magic, Shounen  Special        1   
4235         Action, Adventure, Fantasy, Magic, Shounen    Movie        1   
4445         Action, Adventure, Fantasy, Magic, Shounen    Movie        1   
4978         Action, Adventure, Fantasy, Magic, Shounen  Special        1   
6306         Action, Adventure, Fantasy, Magic, Shounen      OVA        8   
6785         Action, Adventure, Fantasy, Magic, Shounen      OVA        2   
8298         Action, Adventure, Fantasy, Magic, Shounen    Movie        1   
9529         Action, Adventure, Fantasy, Magic, Shounen    Movie        1   
9531         Action, Adventure, Fantasy, Magic, Shounen    Movie        1   
81                                     Military, School    Movie        1   
932                                    Military, School      OVA        1   

      rating  members  
47      8.67    97187  
149     8.41    83056  
83      8.55   197256  
3030    7.10    11447  
101     8.50   245026  
268     8.24   317513  
290     8.22    73480  
461     8.06    52351  
795     7.83   130689  
3833    6.87    10245  
4235    6.77      733  
4445    6.72      736  
4978    6.58    11765  
6306    6.17     2814  
6785    5.93     5372  
8298    6.67     5589  
9529    5.89       71  
9531    5.38       66  
81      8.55    25641  
932     7.76    16143